test/e2e/use-router-with-rewrites/use-router-with-rewrites.test.ts

```diff
@@ -0,0 +1,97 @@
+import { nextTestSetup } from 'e2e-utils'
+
+describe('use-router-with-rewrites', () => {
+  const { next } = nextTestSetup({
+    files: __dirname,
+  })
+
+  it('should preserve current pathname when using useRouter.push with rewrites', async () => {
+    const browser = await next.browser('/')
+    await browser.elementById('router-push').click()
+
+    expect(await browser.url()).toBe(
+      `http://localhost:${next.appPort}/?param=1`
+    )
+  })
+
+  it('should preserve current pathname when using useRouter.replace with rewrites', async () => {
+    const browser = await next.browser('/')
+    await browser.elementById('router-replace').click()
+
+    expect(await browser.url()).toBe(
+      `http://localhost:${next.appPort}/?param=1`
+    )
+  })
+
+  it('should preserve current pathname when using Link with rewrites', async () => {
+    const browser = await next.browser('/')
+    await browser.elementByCss('a').click()
+    expect(await browser.url()).toBe(
+      `http://localhost:${next.appPort}/?param=1`
+    )
+  })
+
+  describe('rewrite to another segment', () => {
+    it('should preserve current pathname when using useRouter.push with rewrites on dynamic route', async () => {
+      const browser = await next.browser('/rewrite-to-another-segment/0')
+      await browser.elementById('router-push').click()
+
+      expect(await browser.url()).toBe(
+        `http://localhost:${next.appPort}/rewrite-to-another-segment/1`
```

Comment:
```suggestion
        `http://localhost:${next.appPort}/rewrite-to-another-segment/0?id=1`
```
This shouldn't merge with the rewrite's query value for `id` it should just append it to the `asPath`'s query so the behavior is correct assertion seems like it needs updating